### PR TITLE
Feat: change profile/group picture

### PR DESCRIPTION
**Pull request by wictornogueira**
# PR Details

## Description

This pr implements methods for updating/removing group icons and client's own profile picture

## Related Issue

- None?

## Motivation and Context

- Just cause

## How Has This Been Tested

```js
const media = await MessageMedia.fromUrl('<url>')
const res = await client.setProfilePicture(media)
```

## Types of changes

- [ ] Dependency change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [X] My code follows the code style of this project.
- [X] I have updated the documentation accordingly (index.d.ts).

### CODE REVIEW

**Comment by alechkos** [comment]
how did you tested the `deleteProfilePicture`?

i tried
```js
client.on('message', async (message) => {
    if (message.body === '!') {
        await client.deleteProfilePicture(message.from);
    }
});
```
and got an error:
```js
my-folder\node_modules\whatsapp-web.js\src\Client.js:1202
        const success = await this.client.pupPage.evaluate((chatid) => {
                                          ^

TypeError: Cannot read properties of undefined (reading 'pupPage')
    at Client.deleteProfilePicture (my-folder\node_modules\whatsapp-web.js\src\Client.js:1202:43)
    at Client.<anonymous> (my-folder\test.js.js:51:22)
    at Client.emit (node:events:513:28)
    at my-folder\node_modules\whatsapp-web.js\src\Client.js:347:18
    at Page._onBindingCalled (my-folder\node_modules\puppeteer\lib\cjs\puppeteer\common\Page.js:998:62)    
    at my-folder\node_modules\puppeteer\lib\cjs\puppeteer\common\Page.js:152:60
    at my-folder\node_modules\puppeteer\lib\cjs\vendor\mitt\src\index.js:51:62
    at Array.map (<anonymous>)
    at Object.emit (my-folder\node_modules\puppeteer\lib\cjs\vendor\mitt\src\index.js:51:43)
    at CDPSession.emit (my-folder\node_modules\puppeteer\lib\cjs\puppeteer\common\EventEmitter.js:72:22)   
    at CDPSession._onMessage (my-folder\node_modules\puppeteer\lib\cjs\puppeteer\common\Connection.js:249:18)
    at Connection._onMessage (my-folder\node_modules\puppeteer\lib\cjs\puppeteer\common\Connection.js:117:25)
    at WebSocket.<anonymous> (my-folder\node_modules\puppeteer\lib\cjs\puppeteer\node\NodeWebSocketTransport.js:13:32)
    at WebSocket.onMessage (my-folder\node_modules\ws\lib\event-target.js:199:18)
    at WebSocket.emit (node:events:513:28)
    at Receiver.receiverOnMessage (my-folder\node_modules\ws\lib\websocket.js:1137:20)
```

**Comment by wictornogueira** [comment]
> how did you tested the `deleteProfilePicture`?
> 
> i tried
> 
> ```js
> client.on('message', async (message) => {
>     if (message.body === '!') {
>         await client.deleteProfilePicture(message.from);
>     }
> });
> ```
> 
> and got an error:
> 
> ```js
> my-folder\node_modules\whatsapp-web.js\src\Client.js:1202
>         const success = await this.client.pupPage.evaluate((chatid) => {
>                                           ^
> 
> TypeError: Cannot read properties of undefined (reading 'pupPage')
>     at Client.deleteProfilePicture (my-folder\node_modules\whatsapp-web.js\src\Client.js:1202:43)
>     at Client.<anonymous> (my-folder\test.js.js:51:22)
>     at Client.emit (node:events:513:28)
>     at my-folder\node_modules\whatsapp-web.js\src\Client.js:347:18
>     at Page._onBindingCalled (my-folder\node_modules\puppeteer\lib\cjs\puppeteer\common\Page.js:998:62)    
>     at my-folder\node_modules\puppeteer\lib\cjs\puppeteer\common\Page.js:152:60
>     at my-folder\node_modules\puppeteer\lib\cjs\vendor\mitt\src\index.js:51:62
>     at Array.map (<anonymous>)
>     at Object.emit (my-folder\node_modules\puppeteer\lib\cjs\vendor\mitt\src\index.js:51:43)
>     at CDPSession.emit (my-folder\node_modules\puppeteer\lib\cjs\puppeteer\common\EventEmitter.js:72:22)   
>     at CDPSession._onMessage (my-folder\node_modules\puppeteer\lib\cjs\puppeteer\common\Connection.js:249:18)
>     at Connection._onMessage (my-folder\node_modules\puppeteer\lib\cjs\puppeteer\common\Connection.js:117:25)
>     at WebSocket.<anonymous> (my-folder\node_modules\puppeteer\lib\cjs\puppeteer\node\NodeWebSocketTransport.js:13:32)
>     at WebSocket.onMessage (my-folder\node_modules\ws\lib\event-target.js:199:18)
>     at WebSocket.emit (node:events:513:28)
>     at Receiver.receiverOnMessage (my-folder\node_modules\ws\lib\websocket.js:1137:20)
> ```

Try:
```typescript
client.on('message', async (message) => {
    if (message.body === '!') {
        const chat = await message.getChat();
        await chat.deletePicture();
    }
});
```

`Client.prototype.deleteProfilePicture` only removes the client's own pfp

**Comment by alechkos** [comment]
> > how did you tested the `deleteProfilePicture`?
> > i tried
> > ```js
> > client.on('message', async (message) => {
> >     if (message.body === '!') {
> >         await client.deleteProfilePicture(message.from);
> >     }
> > });
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and got an error:
> > ```js
> > my-folder\node_modules\whatsapp-web.js\src\Client.js:1202
> >         const success = await this.client.pupPage.evaluate((chatid) => {
> >                                           ^
> > 
> > TypeError: Cannot read properties of undefined (reading 'pupPage')
> >     at Client.deleteProfilePicture (my-folder\node_modules\whatsapp-web.js\src\Client.js:1202:43)
> >     at Client.<anonymous> (my-folder\test.js.js:51:22)
> >     at Client.emit (node:events:513:28)
> >     at my-folder\node_modules\whatsapp-web.js\src\Client.js:347:18
> >     at Page._onBindingCalled (my-folder\node_modules\puppeteer\lib\cjs\puppeteer\common\Page.js:998:62)    
> >     at my-folder\node_modules\puppeteer\lib\cjs\puppeteer\common\Page.js:152:60
> >     at my-folder\node_modules\puppeteer\lib\cjs\vendor\mitt\src\index.js:51:62
> >     at Array.map (<anonymous>)
> >     at Object.emit (my-folder\node_modules\puppeteer\lib\cjs\vendor\mitt\src\index.js:51:43)
> >     at CDPSession.emit (my-folder\node_modules\puppeteer\lib\cjs\puppeteer\common\EventEmitter.js:72:22)   
> >     at CDPSession._onMessage (my-folder\node_modules\puppeteer\lib\cjs\puppeteer\common\Connection.js:249:18)
> >     at Connection._onMessage (my-folder\node_modules\puppeteer\lib\cjs\puppeteer\common\Connection.js:117:25)
> >     at WebSocket.<anonymous> (my-folder\node_modules\puppeteer\lib\cjs\puppeteer\node\NodeWebSocketTransport.js:13:32)
> >     at WebSocket.onMessage (my-folder\node_modules\ws\lib\event-target.js:199:18)
> >     at WebSocket.emit (node:events:513:28)
> >     at Receiver.receiverOnMessage (my-folder\node_modules\ws\lib\websocket.js:1137:20)
> > ```
> 
> Try:
> 
> ```ts
> client.on('message', async (message) => {
>     if (message.body === '!') {
>         const chat = await message.getChat();
>         await chat.deletePicture();
>     }
> });
> ```
> 
> `Client.prototype.deleteProfilePicture` only removes the client's own pfp

tried:
```js
client.on('message', async (message) => {
    if (message.body === '!') {
        const chat = await client.getChatById(message.from);
        await chat.deletePicture();
    }
});
```

an error:
```js
my-folder\node_modules\puppeteer\lib\cjs\puppeteer\common\ExecutionContext.js:221
            throw new Error('Evaluation failed: ' + helper_js_1.helper.getExceptionMessage(exceptionDetails));
                  ^

Error: Evaluation failed: TypeError: window.Store.GroupUtils.requestDeletePicture is not a function
    at Object.window.WWebJS.deletePicture (__puppeteer_evaluation_script__:538:55)
    at __puppeteer_evaluation_script__:2:34
    at ExecutionContext._evaluateInternal (my-folder\node_modules\puppeteer\lib\cjs\puppeteer\common\ExecutionContext.js:221:19)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async ExecutionContext.evaluate (my-folder\node_modules\puppeteer\lib\cjs\puppeteer\common\ExecutionContext.js:110:16)
    at async GroupChat.deletePicture (my-folder\node_modules\whatsapp-web.js\src\structures\GroupChat.js:200:25)
    at async Client.<anonymous> (my-folder\test.js.js:52:9)
```

**Comment by alechkos** [comment]
Perfect, thank you 💯

**Review by group2tts — APPROVED** [review]

**Review by group2tts — APPROVED** [review]

**Comment by CSFelix** [comment]
> It's sick!!!!

File: src/util/Injected.js [hunk 1]
```diff
@@ -625,4 +626,67 @@ exports.LoadUtils = () => {
...
+            return res ? res.status === 200 : false;
+        } catch (err) {
+            if(err.name === 'ServerStatusCodeError') return false;
+            throw err;
+        }
+    };
+
+    window.WWebJS.deletePicture = async (chatid) => {
+        const chatWid = window.Store.WidFactory.createWid(chatid);
+        try {
```
Comment (PurpShell) [inline src/util/Injected.js:684]:
```suggestion
        try {
          const collection = window.Store.ProfilePicThumb.get(chatid);
          if (!collection.canDelete()) return;
```

File: src/util/Injected.js [hunk 2]
```diff
@@ -625,4 +626,67 @@ exports.LoadUtils = () => {
...
+            data: dataUrl.replace(`data:${options.mimeType};base64,`, '')
+        });
+    };
+
+    window.WWebJS.setPicture = async (chatid, media) => {
+        const thumbnail = await window.WWebJS.cropAndResizeImage(media, { asDataUrl: true, mimetype: 'image/jpeg', size: 96 });
+        const profilePic = await window.WWebJS.cropAndResizeImage(media, { asDataUrl: true, mimetype: 'image/jpeg', size: 640 });
+
+        const chatWid = window.Store.WidFactory.createWid(chatid);
+        try {
```
Comment (PurpShell) [inline src/util/Injected.js:673]:
```suggestion
        try {
          const collection = window.Store.ProfilePicThumb.get(chatid);
          if (!collection.canSet()) return;
```

**Review by purpshell — APPROVED** [review]
Good work, here are some changes

File: src/util/Injected.js [hunk 3]
```diff
@@ -625,4 +626,69 @@ exports.LoadUtils = () => {
...
+            if(err.name === 'ServerStatusCodeError') return false;
+            throw err;
+        }
+    };
+
+    window.WWebJS.deletePicture = async (chatid) => {
+        const chatWid = window.Store.WidFactory.createWid(chatid);
+        try {
+          const collection = window.Store.ProfilePicThumb.get(chatid);
+          if (!collection.canDelete()) return;
```
Comment (PurpShell) [inline src/util/Injected.js:686]:
```suggestion
            const collection = window.Store.ProfilePicThumb.get(chatid);
            if (!collection.canDelete()) return;
```

ESLint

**Review by tuyuribr — APPROVED** [review]

**Comment by netomarchiori** [comment]
besides being an admin, what else do I need to check?

I'm trying to change the group chat picture but keep getting `Cannot read properties of undefined (reading 'canSet')` from this piece of code within Injected.js

````
window.WWebJS.setPicture = async (chatid, media) => {
        const thumbnail = await window.WWebJS.cropAndResizeImage(media, { asDataUrl: true, mimetype: 'image/jpeg', size: 96 });
        const profilePic = await window.WWebJS.cropAndResizeImage(media, { asDataUrl: true, mimetype: 'image/jpeg', size: 640 });

        const chatWid = window.Store.WidFactory.createWid(chatid);
        try {
            const collection = window.Store.ProfilePicThumb.get(chatid);
            if (!collection.canSet()) return;

            const res = await window.Store.GroupUtils.sendSetPicture(chatWid, thumbnail, profilePic);
            return res ? res.status === 200 : false;
        } catch (err) {
            if(err.name === 'ServerStatusCodeError') return false;
            throw err;
        }
    };
````

any thoughts?

**Pull request merged**